hera_gpu/redcal.py

```diff
@@ -0,0 +1,819 @@
+    nubls = ubls.shape[1]
+    assert ubls.shape == (ndata, nubls)
+    assert precision in (1,2)
+    assert check_every > 1
+    if verbose:
+        print('PRECISION:', precision)
+        print('NDATA:', ndata)
+        print('NANTS:', nants)
+        print('NUBLS:', nubls)
+        print('NBLS:', nbls)
+
+    # Choose between float/double primitives
+    if precision == 1:
+        real_dtype, complex_dtype = np.float32, np.complex64
+        DTYPE, CDTYPE = 'float', 'cuFloatComplex'
+        CMULT, CONJ, CSUB = 'cuCmulf', 'cuConjf', 'cuCsubf'
+        COPY = cublasCcopy
+    else:
+        real_dtype, complex_dtype = np.float64, np.complex128
+        DTYPE, CDTYPE = 'double', 'cuDoubleComplex'
+        CMULT, CONJ, CSUB = 'cuCmul', 'cuConj', 'cuCsub'
+        COPY = cublasZcopy
+
+    # ensure data types
+    ggu_indices = ggu_indices.astype(np.uint32)
+    data = data.astype(complex_dtype)
+    wgts = wgts.astype(real_dtype)
+    gains = gains.astype(complex_dtype)
+    ubls = ubls.astype(complex_dtype)
+    
+    # px=64, nants=350, precision=1, nthreads//16 -> 15.5 s
+    # px=128, nants=350, precision=1, nthreads//16 -> 15.3 s
+    # px=256, nants=350, precision=1, nthreads//16 -> 16.3 s
+    # px=512, nants=350, precision=1, nthreads//16 -> 18.1 s
+    # px=1024, nants=350, precision=1, nthreads//16 -> 21.5 s
+    # px=2048, nants=350, precision=1, nthreads//16 -> 28.0 s
+    # (killed after that for memory)
+    # Model: 15.3 s + ndata/140 => px=8192 in 75 s on Quadro T2000
+    
+    # px=2000, nants=126, precision=2
+    #chunk_size = min(nthreads // 2, ndata) # 8.5 s
+    #chunk_size = min(nthreads // 4, ndata) # 6.7 s
+    #chunk_size = min(nthreads // 8, ndata) # 5.6 s
+    #chunk_size = min(nthreads // 16, ndata) # 5.2 s
+    #chunk_size = min(nthreads // 32, ndata) # 5.4 s
+    #chunk_size = min(nthreads // 64, ndata) # 5.9 s
+    # px=2000, nants=54, precision=2
+    #chunk_size = min(nthreads // 2, ndata) # 1.5 s
+    #chunk_size = min(nthreads // 4, ndata) # 1.3 s
+    #chunk_size = min(nthreads // 8, ndata) # 1.3 s
+    #chunk_size = min(nthreads // 16, ndata) # 1.5 s
+    #chunk_size = min(nthreads // 32, ndata) # 1.9 s
+    #chunk_size = min(nthreads // 64, ndata) # 3.0 s
+    # px=2000, nants=180, precision=2
+    #chunk_size = min(nthreads // 2, ndata) # 16.8 s
+    #chunk_size = min(nthreads // 4, ndata) # 13.8 s
+    #chunk_size = min(nthreads // 8, ndata) # 11.1 s
+    #chunk_size = min(nthreads // 16, ndata) # 9.9 s
+    #chunk_size = min(nthreads // 32, ndata) # 10.0 s
+    #chunk_size = min(nthreads // 64, ndata) # 10.14 s
+    # px=2000, nants=180, precision=1
+    #chunk_size = min(nthreads // 2, ndata) # 7.6 s
+    #chunk_size = min(nthreads // 4, ndata) # 8.3 s
+    #chunk_size = min(nthreads // 8, ndata) # 7.4 s
+    #chunk_size = min(nthreads // 16, ndata) # 7.3 s
+    #chunk_size = min(nthreads // 32, ndata) # 7.4 s
+    #chunk_size = min(nthreads // 64, ndata) # 7.5 s
```

Comment:
This was info used to choose an optimal chunk size internal to the code. Since it essentially adds traceability to a coding decision, I'd like to leave it in.